torch/_dynamo/polyfills/heapq.py

```diff
@@ -0,0 +1,121 @@
+"""
+Python polyfills for heapq
+"""
+
+from __future__ import annotations
+
+import heapq
+import importlib
+import sys
+from typing import TypeVar
+
+from ..decorators import substitute_in_graph
+
+
+_T = TypeVar("_T")
+
+
+# Partially copied from CPython test/support/import_helper.py
+# https://github.com/python/cpython/blob/bb8791c0b75b5970d109e5557bfcca8a578a02af/Lib/test/support/import_helper.py
+def _save_and_remove_modules(names):  # type: ignore[no-untyped-def]
+    orig_modules = {}
+    prefixes = tuple(name + "." for name in names)
+    for modname in list(sys.modules):
+        if modname in names or modname.startswith(prefixes):
+            orig_modules[modname] = sys.modules.pop(modname)
+    return orig_modules
+
+
+def import_fresh_module(  # type: ignore[no-untyped-def]
+    name, blocked=(), *, deprecated=False, usefrozen=False
+):
+    # Keep track of modules saved for later restoration as well
+    # as those which just need a blocking entry removed
+    blocked = list(blocked)
+    names = {name, *blocked}
+    orig_modules = _save_and_remove_modules(names)
+    for modname in blocked:
+        sys.modules[modname] = None  # type: ignore[assignment]
+
+    try:
+        return importlib.import_module(name)
+    finally:
+        _save_and_remove_modules(names)
+        sys.modules.update(orig_modules)
+
+
+py_heapq = import_fresh_module("heapq", blocked=["_heapq"])
+
+
+__all__ = [
+    "_heapify_max",
+    "heapify",
+    "heappop",
+    "heappush",
+    "heappushpop",
+    "heapreplace",
+    "merge",
+    "nlargest",
+    "nsmallest",
+]
+
+
+@substitute_in_graph(heapq._heapify_max)
+def _heapify_max(heap: list[_T], /) -> None:
+    return py_heapq._heapify_max(heap)
+
+
+if hasattr(py_heapq, "_heappop_max"):
```

Comment:
Assuming this check is related to the changes to this this method across python versions - can we make that check explicit by adding conditionals on the python version as opposed to using has_attr here?